code/game/turfs/simulated/footsteps.dm

```diff
@@ -108,37 +110,48 @@
 	if(buckled || lying || throwing)
 		return //people flying, lying down or sitting do not step
 
-	if(MOVING_QUICKLY(src))
-		if(step_count % 2) //every other turf makes a sound
-			return
+
 
 	if(species.silent_steps)
 		return //species is silent
 
 	if(shoes && (shoes.item_flags & ITEM_FLAG_SILENT))
 		return // quiet shoes
 
-	if(!has_gravity(src))
-		if(step_count % 3) // don't need to step as often when you hop around
-			return
-
-	if(!has_organ(BP_L_FOOT) && !has_organ(BP_R_FOOT))
-		return //no feet no footsteps
+	if(step_count % step_interval)
+		return
 
+	//Nanako's Priority based footstep system.
+	//All structures, floors, and species have a step priority var.
+	//When determining which one gets to choose the footstep sound, highest value wins.
+	//In case of equal values, species > structure > floor
+	//Also heavily optimised an awful structure loop here
 	var/footsound
-	for(var/obj/structure/S in T)
-		footsound = S.get_footstep_sound()
-		if(footsound) break
-	if(!footsound)
+	var/best_priority = 0
+
+	if (species.step_priority > best_priority)
+
+		footsound = get_species_audio(SOUND_FOOTSTEP)
+		best_priority = species.step_priority
+
+	for(var/obj/structure/S as anything in T.step_structures)
```

Comment:
Not the biggest fan of this if I'm honest. A for loop every time you walk sounds expensive. Could you not use something like a step trigger/have a component on the floor which generates the priority step sound at gen / destruction?

Reply:
Have a look at how the code was before. I've massively optimised this loop by creating a new list solely to hold things that affect footsteps. The "as anything" is also a farther optimisation i learned from consulting the byond coding gurus. 

This is massively less efficient than it used to be, and about as good as it can get